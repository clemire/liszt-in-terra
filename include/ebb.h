// The MIT License (MIT)
// 
// Copyright (c) 2016 Stanford University.
// All rights reserved.
// 
// Permission is hereby granted, free of charge, to any person obtaining a
// copy of this software and associated documentation files (the "Software"),
// to deal in the Software without restriction, including without limitation
// the rights to use, copy, modify, merge, publish, distribute, sublicense,
// and/or sell copies of the Software, and to permit persons to whom the
// Software is furnished to do so, subject to the following conditions:
// 
// The above copyright notice and this permission notice shall be included
// in all copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING 
// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
// DEALINGS IN THE SOFTWARE.

#ifndef __EBB_H
#define __EBB_H

#include "terra/terra.h"

#ifdef __APPLE__
#include <sys/syslimits.h>
#define _EBB_H_MAX_PATH_LEN PATH_MAX
#else // assuming Linux here
#include <linux/limits.h>
#define _EBB_H_MAX_PATH_LEN PATH_MAX
#endif

#include <unistd.h>

struct EbbOptions {
  const char *  include_dir;  // path to the ebb include directory
  int           use_gpu;      // 0 if false, 1 if true
};

static inline
int setupebb(lua_State * L, EbbOptions * E) {
    // for composing strings
    char buffer[_EBB_H_MAX_PATH_LEN]; // plenty of room
    size_t bufsize = _EBB_H_MAX_PATH_LEN;

    // check that the Ebb include directory exists at the provided
    // path by looking for this file
    snprintf(buffer, bufsize, "%s/ebb.h", E->include_dir);
    if(access(buffer, R_OK) != 0) {
      fprintf(stderr, "Could not find Ebb include directory at:\n  %s\n",
                      E->include_dir);
      return 1;
    }
    // Then go ahead and modify the terra path so it can find the
    // compiler files
    snprintf(buffer, bufsize,
      "package.terrapath = package.terrapath..';%s/?.t;'",
      E->include_dir);
    if (terra_dostring(L, buffer)) {
        return 1;
    }

    // signal to the compiler that we want to use the GPU
    if (E->use_gpu) {
        lua_pushboolean(L, true);
        lua_setglobal(L, "EBB_USE_GPU_SIGNAL");
    }
    return 0;
}

#endif // __EBB_H
