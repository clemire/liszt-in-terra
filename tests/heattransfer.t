--DISABLE-DISTRIBUTED
-- The MIT License (MIT)
-- 
-- Copyright (c) 2015 Stanford University.
-- All rights reserved.
-- 
-- Permission is hereby granted, free of charge, to any person obtaining a
-- copy of this software and associated documentation files (the "Software"),
-- to deal in the Software without restriction, including without limitation
-- the rights to use, copy, modify, merge, publish, distribute, sublicense,
-- and/or sell copies of the Software, and to permit persons to whom the
-- Software is furnished to do so, subject to the following conditions:
-- 
-- The above copyright notice and this permission notice shall be included
-- in all copies or substantial portions of the Software.
-- 
-- THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
-- IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
-- FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
-- AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
-- LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING 
-- FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
-- DEALINGS IN THE SOFTWARE.
import 'ebb'
local L = require 'ebblib'

local M     = {
  vertices  = L.NewRelation { name="vertices",  size=8  },
  edges     = L.NewRelation { name="edges",     size=12 },
}
M.edges:NewField('head', M.vertices):Load({
  1, 2, 3, 0,   5, 6, 7, 4,   5, 0, 6, 7
})
M.edges:NewField('tail', M.vertices):Load({
  0, 1, 2, 3,   4, 5, 6, 7,   1, 4, 2, 3
})
M.vertices:NewField('position', L.vec3d):Load({
  {0, 0, 0},
  {1, 0, 0},
  {1, 1, 0},
  {0, 1, 0},
  {0, 0, 1},
  {1, 0, 1},
  {1, 1, 1},
  {0, 1, 1},
})

local init_to_zero = terra (mem : &float, i : int) mem[0] = 0 end
local function init_temp (i)
  if i == 0 then
    return 1000
  else
    return 0
  end
end

M.vertices:NewField('flux',        L.float):Load(0)
M.vertices:NewField('jacobistep',  L.float):Load(0)
M.vertices:NewField('temperature', L.float):Load(init_temp)

local ebb compute_step (e : M.edges)
  var v1   = e.head
  var v2   = e.tail
  var dp   = L.vec3f(v1.position - v2.position)
  var dt   = v1.temperature - v2.temperature
  var step = 1.0f / L.length(dp)

  v1.flux += -dt * step
  v2.flux +=  dt * step

  v1.jacobistep += step
  v2.jacobistep += step
end

local ebb propagate_temp (p : M.vertices)
  p.temperature += L.float(.01) * p.flux / p.jacobistep
end

local ebb clear (p : M.vertices)
  p.flux       = 0
  p.jacobistep = 0
end

for i = 1, 1000 do
  M.edges:foreach(compute_step)
  M.vertices:foreach(propagate_temp)
  M.vertices:foreach(clear)
end

M.vertices.temperature:Print()
