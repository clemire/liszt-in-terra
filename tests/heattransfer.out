temperature: <float>
	0	125.46730804443
	1	125.15542602539
	2	125.15542602539
	3	124.8443145752
	4	125.15542602539
	5	124.8443145752
	6	124.8443145752
	7	124.53391265869
